apiserver/facades/controller/charmrevisionupdater/updater.go

```diff
@@ -34,24 +33,50 @@ type CharmRevisionUpdater interface {
 // CharmRevisionUpdaterAPI implements the CharmRevisionUpdater interface and is the concrete
 // implementation of the api end point.
 type CharmRevisionUpdaterAPI struct {
-	state      *state.State
-	resources  facade.Resources
-	authorizer facade.Authorizer
+	state State
+
+	newCharmstoreClient newCharmstoreClientFunc
+	newCharmhubClient   newCharmhubClientFunc
 }
 
+type newCharmstoreClientFunc func(st State) (charmstore.Client, error)
+type newCharmhubClientFunc func(st State, metadata map[string]string) (CharmhubRefreshClient, error)
+
 var _ CharmRevisionUpdater = (*CharmRevisionUpdaterAPI)(nil)
 
 // NewCharmRevisionUpdaterAPI creates a new server-side charmrevisionupdater API end point.
-func NewCharmRevisionUpdaterAPI(
-	st *state.State,
-	resources facade.Resources,
-	authorizer facade.Authorizer,
-) (*CharmRevisionUpdaterAPI, error) {
-	if !authorizer.AuthController() {
+func NewCharmRevisionUpdaterAPI(ctx facade.Context) (*CharmRevisionUpdaterAPI, error) {
+	if !ctx.Auth().AuthController() {
 		return nil, apiservererrors.ErrPerm
 	}
+	api := &CharmRevisionUpdaterAPI{
```

Comment:
Call NewCharmRevisionUpdaterAPIState here, rather than return the struct.

Reply:
Done